### PR TITLE
feat: workspace CRUD and workspace switcher (#26)

### LINKED ISSUE #26: Workspace CRUD and workspace switcher
## Description

Allow users to create additional workspaces (up to 2 beyond their personal workspace, 3 total), edit workspace settings (name, slug), switch between workspaces, and delete non-personal workspaces. The personal workspace is always listed first and cannot be deleted.

## Acceptance Criteria

- [ ] User can create a new workspace (name, auto-generated slug)
- [ ] Workspace creation is blocked when user has 3 workspaces (button disabled with limit message)
- [ ] Workspace creation limit enforced server-side (DB trigger rejects insert if count ≥ 3)
- [ ] Workspace switcher in sidebar lists all workspaces (personal first, then others alphabetically)
- [ ] Clicking a workspace in the switcher navigates to `/[workspaceSlug]`
- [ ] Workspace settings page at `/[workspaceSlug]/settings` — edit name and slug
- [ ] User can delete a non-personal workspace (with confirmation dialog)
- [ ] Deleting a workspace cascades to all pages and members
- [ ] Personal workspace cannot be deleted (delete option hidden or disabled with explanation)
- [ ] Workspaces the user was invited to do not count toward their creation limit

## Dependencies

Depends on #25 (App shell — sidebar with workspace switcher placeholder).

## Technical Notes

- Workspace CRUD uses `@supabase/supabase-js` — no ORM
- The DB trigger from #23 enforces the creation limit server-side; also enforce client-side for UX
- Slug generation: lowercase, hyphenated, unique — validate on create/update
- Use shadcn/ui Dialog for create workspace modal, AlertDialog for delete confirmation
- Settings page route: `(app)/[workspaceSlug]/settings/page.tsx`

### PR DESCRIPTION
**Pull request by zacharias-ona**
Closes #26

## What

Workspace CRUD operations and a functional workspace switcher in the sidebar. Users can create additional workspaces (up to 3 total), switch between them, edit workspace settings, and delete non-personal workspaces.

## How

**Workspace switcher** (`workspace-switcher.tsx`): Dropdown in the sidebar listing all workspaces the user belongs to. Personal workspace listed first, others alphabetically. Includes "Create workspace" button (disabled at limit) and "Workspace settings" link for non-personal workspaces.

**Create workspace dialog** (`create-workspace-dialog.tsx`): Dialog with name input, auto-generated slug with random suffix for uniqueness. Creates workspace + owner membership. Client-side limit check disables the button; server-side DB trigger rejects inserts at ≥3.

**Settings page** (`[workspaceSlug]/settings/page.tsx` + `workspace-settings-form.tsx`): Edit workspace name and slug with validation. Delete non-personal workspaces via AlertDialog confirmation — cascades to all pages and members via DB foreign keys. Personal workspace shows explanation that it cannot be deleted.

**Data flow**: `AppShell` loads all workspaces via members join, passes them to sidebar components. `AppLayout` passes `userId` from the auth guard.

## New dependencies

- shadcn/ui `dialog` and `alert-dialog` components (both use `@base-ui/react` primitives already in the project)

## Testing

- Unit tests for `generateSlug`, `isValidSlug`, `MAX_CREATED_WORKSPACES`
- `pnpm lint && pnpm typecheck && pnpm test` — all pass (20 tests)
- Mobile responsive: sidebar Sheet works with updated workspace switcher

### CODE REVIEW

File: src/components/sidebar/create-workspace-dialog.tsx [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+      return;
+    }
+
+    // Create owner membership for the new workspace
+    await supabase.from("members").insert({
+      workspace_id: data.id,
+      user_id: userId,
+      role: "owner",
+      joined_at: new Date().toISOString(),
+    });
```
Comment (zacharias-ona) [inline src/components/sidebar/create-workspace-dialog.tsx:93]: **Bug: Member insert will fail due to RLS.**

The `members` table INSERT policy requires `is_workspace_admin(workspace_id)`, which checks if the current user is already an owner/admin of the workspace. At this point the workspace was just created and has no members yet, so this insert will be rejected by RLS.

Fix: Add a database trigger on `workspaces` that auto-creates an owner membership row when a new workspace is inserted (similar to how `handle_new_user` creates the membership for the personal workspace). This keeps the security boundary in the DB layer and removes the need for this client-side insert entirely.

```sql
create or replace function auto_create_workspace_owner()
returns trigger
language plpgsql
security definer
set search_path = ''
as $$
begin
  insert into public.members (workspace_id, user_id, role, joined_at)
  values (new.id, new.created_by, 'owner', now());
  return new;
end;
$$;

create trigger auto_create_workspace_owner_trigger
  after insert on workspaces
  for each row
  execute function auto_create_workspace_owner();
```

Then remove lines 87–93 from this file.</p>">

File: src/app/(app)/[workspaceSlug]/settings/page.tsx [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  }
+
+  return (
+    <div className="mx-auto max-w-xl p-6">
+      <h1 className="text-2xl font-semibold">Workspace settings</h1>
+      <p className="mt-1 text-sm text-muted-foreground">
+        Manage your workspace name, URL, and other settings.
+      </p>
+      <div className="mt-6">
+        <WorkspaceSettingsForm workspace={workspace as Workspace} />
```
Comment (zacharias-ona) [inline src/app/(app)/[workspaceSlug]/settings/page.tsx:31]: **Convention: Unjustified `as` cast.**

AGENTS.md: *No `as` casts unless unavoidable (add a comment explaining why).*

The Supabase `.select("*")` return type doesn't match `Workspace` exactly, but the cast can be avoided by selecting the specific columns that match the `Workspace` interface:

```typescript
const { data: workspace } = await supabase
  .from("workspaces")
  .select("id, name, slug, is_personal, created_by, created_at, updated_at")
  .eq("slug", workspaceSlug)
  .maybeSingle();
```

If the cast is truly needed due to Supabase's generated types not being configured, add a comment explaining why per convention.">

File: src/app/(app)/[workspaceSlug]/settings/page.tsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  params: Promise<{ workspaceSlug: string }>;
+}) {
+  const { workspaceSlug } = await params;
+  const supabase = await createClient();
+
+  const { data: workspace } = await supabase
+    .from("workspaces")
+    .select("*")
+    .eq("slug", workspaceSlug)
+    .maybeSingle();
```
Comment (zacharias-ona) [inline src/app/(app)/[workspaceSlug]/settings/page.tsx:18]: **Security: No authorization check — any authenticated user can view/edit any workspace's settings.**

The page fetches the workspace by slug without verifying the current user is a member (or admin). The RLS policy on `workspaces` SELECT requires `is_workspace_member(id)`, so the query itself is protected. However, the UPDATE and DELETE operations in `WorkspaceSettingsForm` happen client-side and are protected by RLS there too.

This is actually fine — RLS handles it. Disregard if the intent is to rely on RLS for authorization. But consider: if a non-member navigates to `/{slug}/settings`, they'll see a 404 (because the SELECT returns nothing due to RLS), which is the correct behavior. ✅ No action needed on this one — just confirming the auth model is sound.">

File: src/components/sidebar/app-shell.tsx [hunk 4]
```diff
@@ -1,41 +1,56 @@
...
-      .eq("slug", params.workspaceSlug)
-      .maybeSingle()
+      .from("members")
+      .select("workspace_id, workspaces(*)")
+      .eq("user_id", userId)
       .then(({ data }) => {
-        if (data) setWorkspaceName(data.name);
+        if (data) {
+          const ws = data
+            .map((m) => m.workspaces as unknown as Workspace)
```
Comment (zacharias-ona) [inline src/components/sidebar/app-shell.tsx:36]: **Convention: `as unknown as Workspace` — unjustified double cast.**

AGENTS.md: *No `as` casts unless unavoidable (comment why).*

Supabase's join return type for `workspaces(*)` is a nested object that doesn't directly match `Workspace`. This is a known Supabase typing limitation with joined selects. Add a comment explaining why:

```typescript
// Supabase join types return workspaces as a nested object; cast needed
// because generated types aren't configured for this project yet.
.map((m) => m.workspaces as unknown as Workspace)
```

Same pattern appears in `workspace-settings-form.tsx` line 120 (`as unknown as { slug: string }`). Add justification comments to both.">

**Comment by zacharias-ona** [comment]
## Review Summary

Reviewed against AGENTS.md, `.agents/conventions.md`, `.agents/design.md`, and the linked issue #26.

**Requesting changes** — 1 blocking bug, 1 convention issue.

### Blocking

1. **RLS bug in workspace creation flow** (`create-workspace-dialog.tsx:88-93`): After inserting a new workspace, the code inserts a member row client-side. The `members` INSERT RLS policy requires `is_workspace_admin(workspace_id)`, which checks for an existing owner/admin membership — but none exists yet for the new workspace. This insert will be silently rejected by RLS. Fix: add a DB trigger (`AFTER INSERT ON workspaces`) that auto-creates the owner membership, then remove the client-side member insert.

### Non-blocking (fix while you're in there)

2. **Unjustified `as` casts** (`settings/page.tsx:31`, `app-shell.tsx:36`, `workspace-settings-form.tsx:120`): AGENTS.md requires a comment explaining why for any `as` cast. Either add justification comments or eliminate the casts by selecting explicit columns.

### What looks good

- Workspace switcher UX: personal first, alphabetical sort, limit enforcement both client and server-side ✅
- Design spec compliance: `text-sm` body text, `text-xs` secondary text, `gap-2` icon spacing, sharp corners, correct color tokens ✅
- Settings page: proper `notFound()` for missing workspace, slug validation, delete confirmation with AlertDialog ✅
- Tests cover `generateSlug`, `isValidSlug`, and `MAX_CREATED_WORKSPACES` ✅
- Auth model: RLS policies correctly protect all operations; settings page returns 404 for non-members ✅
- No scope creep — changes are focused on the workspace CRUD feature ✅

File: src/components/sidebar/create-workspace-dialog.tsx (hunk 1 above)
Comment (zacharias-ona) [inline src/components/sidebar/create-workspace-dialog.tsx:93]: **Bug: unhandled error on member insert.** If this insert fails (network error, RLS issue, etc.), the workspace exists but the user has no membership — an orphaned workspace they can't access or delete.

Handle the error and roll back the workspace creation on failure:

```typescript
const { error: memberError } = await supabase.from("members").insert({
  workspace_id: data.id,
  user_id: userId,
  role: "owner",
  joined_at: new Date().toISOString(),
});

if (memberError) {
  // Roll back: delete the workspace we just created
  await supabase.from("workspaces").delete().eq("id", data.id);
  setError("Failed to set up workspace membership. Please try again.");
  setLoading(false);
  return;
}
```

File: src/components/sidebar/app-shell.tsx (hunk 4 above)
Comment (zacharias-ona) [inline src/components/sidebar/app-shell.tsx:36]: **Convention: `as` cast requires a comment.** Per AGENTS.md: "No `as` casts unless unavoidable (comment why)."

Supabase join queries return the joined relation as an opaque type that doesn't match the app's `Workspace` interface, so the double cast is needed. Add a comment:

```typescript
// Supabase join returns workspaces as an opaque type; cast to match app interface
.map((m) => m.workspaces as unknown as Workspace)
```

Same applies to the cast in `workspace-settings-form.tsx` line 138 and `settings/page.tsx` line 31.

**Review by zacharias-ona — COMMENTED** [review]
## Review: Changes Requested

Two issues to address:

1. **Unhandled error on member insert** (create-workspace-dialog.tsx:88-93): The member insert after workspace creation ignores errors. If it fails, the workspace is orphaned — the user can't access or delete it. Handle the error and roll back the workspace insert on failure.

2. **`as` casts missing required comments** (app-shell.tsx:36, workspace-settings-form.tsx:138, settings/page.tsx:31): AGENTS.md requires a comment explaining why each `as` cast is unavoidable. Add a brief comment at each cast site.

Everything else looks good — RLS policies cover auth, the workspace switcher sorts correctly, slug generation and validation are well-tested, and the UI follows the design spec.

File: src/components/sidebar/create-workspace-dialog.tsx (hunk 1 above)
Comment (zacharias-ona) [inline src/components/sidebar/create-workspace-dialog.tsx:93]: **Bug: unhandled error on member insert orphans the workspace.**

If this insert fails, the workspace exists but the user has no membership — they can't access or delete it. The workspace is orphaned.

Handle the error and clean up the workspace on failure:

```typescript
const { error: memberError } = await supabase.from("members").insert({
  workspace_id: data.id,
  user_id: userId,
  role: "owner",
  joined_at: new Date().toISOString(),
});

if (memberError) {
  // Roll back: delete the orphaned workspace
  await supabase.from("workspaces").delete().eq("id", data.id);
  setError("Failed to set up workspace membership. Please try again.");
  setLoading(false);
  return;
}
```

File: src/components/sidebar/app-shell.tsx (hunk 4 above)
Comment (zacharias-ona) [inline src/components/sidebar/app-shell.tsx:36]: **Convention: `as` cast missing required comment.**

AGENTS.md: "No `as` casts unless unavoidable (comment why)."

Add a comment explaining why the cast is needed:

```typescript
// Supabase join returns workspaces as a nested object type; cast to our Workspace interface
.map((m) => m.workspaces as unknown as Workspace)
```

File: src/components/workspace/workspace-settings-form.tsx [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+    if (user) {
+      const { data: membership } = await supabase
+        .from("members")
+        .select("workspace_id, workspaces(slug)")
+        .eq("user_id", user.id)
+        .limit(1)
+        .maybeSingle();
+
+      if (membership?.workspaces) {
+        const ws = membership.workspaces as unknown as { slug: string };
```
Comment (zacharias-ona) [inline src/components/workspace/workspace-settings-form.tsx:138]: **Convention: `as` cast missing required comment.**

Same issue — add a comment explaining the cast:

```typescript
// Supabase join returns workspaces as a nested object; cast to access slug
const ws = membership.workspaces as unknown as { slug: string };
```

File: src/app/(app)/[workspaceSlug]/settings/page.tsx (hunk 2 above)
Comment (zacharias-ona) [inline src/app/(app)/[workspaceSlug]/settings/page.tsx:31]: **Convention: `as` cast missing required comment.**

Add a comment explaining why:

```typescript
{/* Supabase .select("*") returns a generic row type; cast to Workspace since the query shape matches */}
<WorkspaceSettingsForm workspace={workspace as Workspace} />
```

**Comment by zacharias-ona** [comment]
## Review: Changes Requested

Two issues to address:

1. **Unhandled error on member insert** (`create-workspace-dialog.tsx:88-93`): The member insert after workspace creation ignores errors. If it fails, the workspace is orphaned — the user can't access or delete it. Handle the error and roll back the workspace insert on failure.

2. **`as` casts missing required comments** (`app-shell.tsx:36`, `workspace-settings-form.tsx:138`, `settings/page.tsx:31`): AGENTS.md requires a comment explaining why each `as` cast is unavoidable. Add a brief comment at each cast site.

Everything else looks good — RLS policies cover auth, the workspace switcher sorts correctly, slug generation and validation are well-tested, and the UI follows the design spec.

File: src/components/sidebar/create-workspace-dialog.tsx (hunk 1 above)
Comment (zacharias-ona) [inline src/components/sidebar/create-workspace-dialog.tsx:93]: **Bug: Unhandled member insert error orphans the workspace.**

If this insert fails (network error, RLS rejection, etc.), the workspace exists but the user has no membership — they can't access or delete it.

Additionally, the `members` INSERT RLS policy requires `is_workspace_admin(workspace_id)`, which checks if the caller is already an owner/admin. Since the workspace was just created with no members, this insert will be rejected by RLS.

**Fix:** Create a database trigger that auto-creates an owner membership when a workspace is inserted (similar to how `handle_new_user` creates the personal workspace membership). Then remove lines 87–93 entirely.

```sql
create or replace function auto_create_workspace_owner()
returns trigger language plpgsql security definer set search_path = ''
as $$
begin
  insert into public.members (workspace_id, user_id, role, joined_at)
  values (new.id, new.created_by, 'owner', now());
  return new;
end;
$$;

create trigger auto_create_workspace_owner_trigger
  after insert on workspaces for each row
  execute function auto_create_workspace_owner();
```

This keeps the security boundary in the DB layer and eliminates the race condition.

File: src/components/sidebar/app-shell.tsx (hunk 4 above)
Comment (zacharias-ona) [inline src/components/sidebar/app-shell.tsx:36]: **Convention: `as unknown as Workspace` cast missing required comment.**

AGENTS.md: "No `as` casts unless unavoidable (add a comment explaining why)."

Add a comment explaining the cast:

```typescript
// Supabase join returns workspaces as an opaque nested type; cast to match app Workspace interface
.map((m) => m.workspaces as unknown as Workspace)
```

File: src/app/(app)/[workspaceSlug]/settings/page.tsx (hunk 2 above)
Comment (zacharias-ona) [inline src/app/(app)/[workspaceSlug]/settings/page.tsx:31]: **Convention: `as Workspace` cast missing required comment.**

AGENTS.md: "No `as` casts unless unavoidable (add a comment explaining why)."

Either add a justification comment, or avoid the cast by selecting specific columns:

```typescript
const { data: workspace } = await supabase
  .from("workspaces")
  .select("id, name, slug, is_personal, created_by, created_at, updated_at")
  .eq("slug", workspaceSlug)
  .maybeSingle();
```

File: src/components/workspace/workspace-settings-form.tsx [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+    const supabase = createClient();
+
+    const { error: deleteError } = await supabase
+      .from("workspaces")
+      .delete()
+      .eq("id", workspace.id);
+
+    if (deleteError) {
+      setError(deleteError.message);
+      setDeleting(false);
```
Comment (zacharias-ona) [inline src/components/workspace/workspace-settings-form.tsx:120]: **Convention: `as unknown as { slug: string }` cast missing required comment.**

Same issue — add a comment explaining why the cast is needed:

```typescript
// Supabase join returns workspaces as an opaque nested type; cast to access slug
const ws = membership.workspaces as unknown as { slug: string };
```

**Comment by zacharias-ona** [comment]
## Review: Changes Requested

### Blocking

1. **Member insert fails due to RLS / orphans workspace on error** (`create-workspace-dialog.tsx:87-93`): The client-side member insert after workspace creation will be rejected by the `members` INSERT RLS policy (user isn't yet an admin of the new workspace). Even if it weren't blocked by RLS, the error is silently ignored — a failure would leave an orphaned workspace. Fix: add a DB trigger to auto-create the owner membership on workspace insert, and remove the client-side insert.

2. **`as` casts missing required comments** (`app-shell.tsx:36`, `settings/page.tsx:31`, `workspace-settings-form.tsx:120`): AGENTS.md requires a comment explaining why each `as` cast is unavoidable. Three cast sites need justification comments added.

### Non-blocking note

3. **Merge conflict**: This PR has conflicts with `main` (likely from PR #48 merging). Needs a rebase before merge.

### What looks good

- Workspace switcher sorts correctly (personal first, then alphabetical) ✅
- Client-side limit check + server-side DB trigger for workspace creation limit ✅
- Slug generation and validation are well-tested (unit tests cover edge cases) ✅
- Settings page auth is handled by RLS — non-members get 404 ✅
- Delete cascade via DB foreign keys is correct ✅
- UI follows design spec (typography, spacing, color tokens) ✅
- shadcn/ui Dialog and AlertDialog components are standard generated code ✅

**Comment by zacharias-ona** [comment]
Closing as duplicate. PR #46 implemented the same feature (issue #26) and was already merged. This PR was created by a race condition between the Feature Builder's webhook and cron triggers firing simultaneously — fixed in PR #50.
